### ISSUE TITLE
Inconsistent type evaluation with `Self`

### ISSUE THREAD

**Issue by Azureblade3808** [microsoft/pyright#4844]
**Describe the bug**
It seems sometimes `Self` gets narrowed too much, but I'm not sure about the condition.

**To Reproduce**
See "Screenshots or Code".

**Expected behavior**
Nothing gets reported.

**Screenshots or Code**
```python
from __future__ import annotations

from typing import Protocol
from typing_extensions import Self

class Negatable(Protocol):
    def __neg__(self) -> Self: ...

def f(x: Negatable) -> None: ...

f(0)  # This fails.
f(int(0))
f(0)  # This passes.
```

![image](https://user-images.githubusercontent.com/15389744/227828194-ab25c2ad-7a6f-4a1f-873e-a5a29b908ee5.png)

**VS Code extension or command-line**
Pylance v2023.3.31 Pre-Release

**Additional context**
VSCode config:
```json
"python.analysis.diagnosticSeverityOverrides": {
    "reportAssertAlwaysTrue": "error",
    "reportConstantRedefinition": "error",
    "reportDuplicateImport": "warning",
    "reportFunctionMemberAccess": "error",
    "reportGeneralTypeIssues": "information",
    "reportImplicitStringConcatenation": "warning",
    "reportImportCycles": "none",
    "reportIncompatibleMethodOverride": "error",
    "reportIncompatibleVariableOverride": "error",
    "reportInconsistentConstructor": "warning",
    "reportInvalidStubStatement": "none",
    "reportMissingModuleSource": "none",
    "reportMissingTypeArgument": "none",
    "reportMissingTypeStubs": "none",
    "reportOverlappingOverload": "none",
    "reportPrivateImportUsage": "none",
    "reportPrivateUsage": "none",
    "reportPropertyTypeMismatch": "none",
    "reportUnknownArgumentType": "none",
    "reportUnknownLambdaType": "none",
    "reportUnknownMemberType": "none",
    "reportUnknownParameterType": "none",
    "reportUnknownVariableType": "none",
    "reportUnnecessaryCast": "none",
    "reportUnnecessaryComparison": "none",
    "reportUnnecessaryIsInstance": "none",
    "reportUnsupportedDunderAll": "warning",
    "reportUntypedClassDecorator": "none",
    "reportUntypedFunctionDecorator": "none",
    "reportUnusedCallResult": "error",
    "reportUnusedClass": "error",
    "reportUnusedFunction": "error",
    "reportUnusedImport": "warning",
    "reportUnusedVariable": "error"
},
"python.analysis.typeCheckingMode": "strict",
```

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.301, which I just published. It will also be included in a future release of pylance.
